add in-app and system notifications

oshorefueled opened the pull request:
This PR resolves #527 

It implements and in-app notification (toast) which is used in displaying success or error messages to a user within the app in response to an action. It also implements a system notification which displays os specific system-wide applications. Some of the notifications displayed with the system notification could be new block notifications, receive transaction notifications, transaction confirmation notification e.t.c..

Calls to the CreateToast method used by pages in the page package are replaced with the Notify method created in this PR.

oshorefueled committed:
add notification package

- add toast nofication implementation to notification page
- move toast layout code to toast.go
- add toast to load and common structs
- remove previous toast implementation

oshorefueled committed:
implement system notification

oshorefueled committed:
stack toast layout on modals layout

oshorefueled committed:
go import log file

oshorefueled committed:
resolve build error from undeclared toast variable

beansgum commented on ui/notification/toast.go:
```diff
@@ -0,0 +1,77 @@
+	t.message = message
+	t.success = success
+	t.timer = time.NewTimer(time.Second * 3)
+	t.show = true
```
Protect these variables with a mutex.

beansgum commented on ui/notification/toast.go:
```diff
@@ -0,0 +1,77 @@
+}
+
+func (t *Toast) Layout(gtx layout.Context) layout.Dimensions {
+	if !t.show {
```
Remove the show variable and replace this condition with `t.timer == nil`.

beansgum commented on ui/notification/toast.go:
```diff
@@ -0,0 +1,77 @@
+		return layout.Dimensions{}
+	}
+
+	t.handleToastDisplay(gtx)
```
This check should be done at the top of the function.

oshorefueled replied:
moving the handle function to the top of the Layout method would lead to a nil dereference error.

beansgum commented on ui/notification/log.go:
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) 2017, The dcrdata developers
```
This package does not need its own logger.

beansgum commented on ui/notification/system.go:
```diff
@@ -0,0 +1,35 @@
+func (s *SystemNotification) Notify(message string) {
+	err := beeep.Notify(title, message, s.iconPath)
+	if err != nil {
+		log.Info("could not initiate desktop notification, reason:", err.Error())
```
This error should be returned not logged.

oshorefueled replied:
The log avoids handling notification errors in the ui code. All that needs to be called is the Notify method, just like toast notifications.

beansgum commented on ui/notification/utils.go:
```diff
@@ -0,0 +1,22 @@
+	"path/filepath"
+)
+
+func getAbsolutePath() (string, error) {
```
I don't think any utils.go file is needed in this package since it's just this function and it's used only in system.go

oshorefueled committed:
add mutex to toast struct

oshorefueled committed:
add delay speeds to toast notification

oshorefueled committed:
resolve undefined variable causing build to fail

oshorefueled committed:
change variable name from delay to duration

oshorefueled committed:
refactor notification method

- add NotifyError method
- add comments to Notify methods
- call NotifyError for error messages in layout code and Notify for success
  messages.

oshorefueled committed:
go import send page

Sirmorrison reviewed (approved):
Tested and looks good.
